blockchain/blockchain_test.go

```diff
@@ -1644,3 +1658,403 @@ func TestBlockchain_WriteFullBlock(t *testing.T) {
+}
+
+func blockWriter(tb testing.TB, numberOfBlocks uint64, blockTime, checkInterval time.Duration, byteToRead []byte, receiptsBytesToRead []byte) {
+	tb.Helper()
+
+	blockTicker := time.NewTicker(blockTime)
+
+	type blockCounter struct {
+		mu sync.RWMutex
+		x  uint64
+	}
+
+	counter := &blockCounter{x: 0}
+
+	quitChan := make(chan bool)
+
+	p := fmt.Sprintf("/tmp/blockchain-disk-usage-test")
+	err := os.Mkdir(p, 0775)
+	require.NoError(tb, err)
+
+	tb.Cleanup(func() {
+		if err := os.RemoveAll(p); err != nil {
+			tb.Fatal(err)
+		}
+	})
+
+	db, err := leveldb.NewLevelDBStorage(
+		filepath.Join(p),
+		hclog.NewNullLogger(),
+	)
+	require.NoError(tb, err)
+
+	block, err := customJSONBlockUnmarshall(tb, byteToRead)
+	require.NoError(tb, err)
+
+	receipts, err := customJSONReceiptsUnmarshall(tb, receiptsBytesToRead)
+	require.NoError(tb, err)
+
+	dirSizeCheck := func() {
+		ticker := time.NewTicker(checkInterval)
+
+		for {
+			select {
+			case <-quitChan:
+				return
+			case <-ticker.C:
+				dirSizeValue, err := dirSize(tb, p)
+
+				if err != nil {
+					tb.Log(err)
+				}
+
+				counter.mu.RLock()
+
+				tb.Logf("BLOCK: %d DIRSIZE IS: %d and average is:%.2f", counter.x, dirSizeValue, float64(dirSizeValue)/float64(counter.x))
+
+				counter.mu.RUnlock()
+			}
+		}
+	}
+
+	blockchain := &Blockchain{db: db}
+
+	dirSizeBeforeBlocks, err := dirSize(tb, p)
+	require.NoError(tb, err)
+	tb.Logf("DIRSIZE IS: %d", dirSizeBeforeBlocks)
+
+	go dirSizeCheck()
+
+	for ; ; <-blockTicker.C {
+		batchWriter := storage.NewBatchWriter(db)
+
+		counter.mu.RLock()
+		block.Block.Header.Number = counter.x
+		block.Block.Header.Hash = types.StringToHash(fmt.Sprintf("%d", counter.x))
+
+		for i, transaction := range block.Block.Transactions {
+			transaction.Nonce = counter.x * uint64(i)
+			addr := types.StringToAddress(fmt.Sprintf("%d", counter.x*uint64(i)))
+			transaction.To = &addr
+		}
+
+		counter.mu.RUnlock()
+
+		batchWriter.PutHeader(block.Block.Header)
+		batchWriter.PutBody(block.Block.Hash(), block.Block.Body())
+
+		batchWriter.PutReceipts(block.Block.Hash(), receipts)
+
+		require.NoError(tb, blockchain.writeBatchAndUpdate(batchWriter, block.Block.Header, big.NewInt(0), false))
+
+		counter.mu.Lock()
+
+		counter.x++
+
+		counter.mu.Unlock()
+
+		counter.mu.RLock()
+
+		if counter.x == numberOfBlocks {
+			counter.mu.RUnlock()
+
+			break
+		}
+
+		counter.mu.RUnlock()
+	}
+
+	quitChan <- true
+
+	dirSizeAfterBlocks, err := dirSize(tb, p)
+	require.NoError(tb, err)
+	tb.Logf("DIRSIZE After All blocks: %d, average blocks size: %.2f", dirSizeAfterBlocks, float64(dirSizeAfterBlocks)/float64(numberOfBlocks))
+
+	db.Close()
+
+	assert.NotEqual(tb, dirSizeBeforeBlocks, dirSizeAfterBlocks)
+}
+
+func customJSONBlockUnmarshall(tb testing.TB, jsonData []byte) (*types.FullBlock, error) {
+	tb.Helper()
+
+	var ( //nolint:prealloc
+		dat          map[string]interface{}
+		err          error
+		transactions []*types.Transaction
```

Comment:
You could have provided initial capacity to `transactions` slice, but as you wish.